### DISCUSSION TITLE
How to tune flow deck state estimation

### DISCUSSION THREAD

**Discussion by SharyatSB** [bitcraze/discussions#958] [Q&A]
Hey all, I am working on a mapping project with a crazyflie drone. I am using the multiranger for this project. The basic idea is that while collecting points, the drone turns 90 degrees and collects the distance from all the TOF sensors and also the yaw of the drone. After that, I post process the data and converted it into x-y coordinates. Here is the map with just the multiranger data. 

![MAP_local](https://github.com/bitcraze/discussions/assets/77988600/01d0e3ae-87e6-4016-a45d-0e3e488eb8b6)

I understand that mapping with my method is not the most ideal approach and I am open to any feedback. 
Note: In the previous approach, the model assumes that the drone is always at the origin of the global frame. 

Next, my plan was to use state estimation in x and y to improve the quality of the map and this would also help with creating a map while exploring an environment when the drone is not at the origin of the global frame. I basically added the x state estimate with the x data from the previous approach and same for y. However, my map actually got worse. Here is a screenshot:

![MAP_Global](https://github.com/bitcraze/discussions/assets/77988600/3b77e88f-582e-4803-ae49-9c54da8a9e0f)

here, blue is the position of the drone over time as calculated by CF  state estimation. 

I was hoping if I could get some suggestions on how I can make the state estimation more accurate? From what I understand, the cf calculates the state estimation from the flow deck which requires some patterns to accurately detect the changes in frames. In this case, I put a poster with some letters and drawings under the drone to help the flow deck. Not sure if this actually worked or made it worse?

Another approach that I could try is setting the state estimate as zero once it has taken off and levelled because I think majority of the deviation happens when the drone takes off. 

Also, I know is is a very basic approach to mapping and I hope to advance to much more sophisticated techniques but I think is a good place to start. I would also appreciate any resources that might help with my knowledge of aerial SLAM. 

Thanks!

**Comment by knmcguire** [comment]
Hi!

Let me give you some links to get started
- The multiranger pointcloud example, which does what you are trying to do but in 3d: https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/multiranger/multiranger_pointcloud.py
- The crazyswarm2 project with ROS examples https://imrclab.github.io/crazyswarm2/tutorials.html#mapping-with-the-slam-toolbox
- I did a summer mapping project with with the multiranger last summer with ROS: https://github.com/knmcguire/crazyflie_ros2_experimental

Here are also some recent slam papers as well.
- Efficient 2D Graph SLAM for Sparse Sensing. Hanzhi Zhou, Zichao Hu, Sihang Liu, and Samira Khan. In 2022 IEEE/RSJ International Conference on Intelligent Robots and Systems (IROS). 2022
- Niculescu, Vlad, et al. "NanoSLAM: Enabling Fully Onboard SLAM for Tiny Robots." arXiv preprint arXiv:2309.12008 (2023). (this is with an special deck made for the crazyflie)

In terms of improving the map, just mind that the flowdeck can't detect it is rotating. If you have it fly in place and rotate, then drift is more likely to occur. So I would advise to have the crazyflie fly in a small circle instead, and ofcourse make sure it is flying over enough texture.

What you can also do some assumptions on your room and perhaps do a Ransac line estimation of the walls. I've seen that done on an IROS poster work called 'Parsing Indoor Manhattan Scenes Using Four-Point LiDAR on a Micro UAV'.

And last but not least, after my summer project I also found that somebody tried to map with the crazyflie as well way before me :) https://pramodatre.vercel.app/blog/2020-10-16-slam-with-crazyflie

About your suggestion to reset the state estimation. that is probably a good idea to start mapping once you are in the air. But ofcourse don't restart the state estimate itself of the crazyflie itself or else you crash :P. You can monitor your height in the script and just turn on the mapping once you are at 80 cms or something.

All in all, the 'official way' to improve this, is to implement SLAM fully on your example. Currently, it is only doing mapping and with SLAM, the range measurements and state estimation enhance each other over time for self-localization and mapping. But then you probably would need to get into some theory there before starting something like that.

This will probably get you started somewhere :) In terms of support, it is probably good to have one discussion per issue in the future, so that should probably be your approach after you tried some things and still have questions.
